### PR TITLE
Various BuildUpToDateCheck optimisations

### PR DESCRIPTION
**Pull request by drewnoakes**
Best reviewed commit by commit. Commit messages explain my intentions for each change. The combined diff is a bit much in one go.

I'm happy for this to be split up into smaller PRs, but there are dependencies between the commits. Given the time zone differences between us I'd be interested in feedback on the lot.

### CODE REVIEW

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UpToDate/BuildUpToDateCheck.cs [hunk 1]
```diff
@@ -394,7 +394,13 @@ private static (DateTime? time, string path) GetLatestInput(IEnumerable<string>
...
             {
                 DateTime? time = GetTimestamp(input, timestampCache);
-                if (latest != null && (time == null && !ignoreMissing || time > latest))
+
+                if (time == null && !ignoreMissing)
+                {
+                    return (null, null);
+                }
+
+                if (time > latest)
```
Comment (tmeschter) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UpToDate/BuildUpToDateCheck.cs]: I'll say I find this much easier to read.

**Comment by davkean** [comment]
tag @lifengl

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UpToDate/BuildUpToDateCheck.cs [hunk 2]
```diff
@@ -346,10 +346,11 @@ private bool CheckGlobalConditions(BuildAction buildAction, BuildUpToDateCheckLo
...
             }
 
-            (string Path, string Link, CopyToOutputDirectoryType CopyType) copyAlwaysItem = _items.SelectMany(kvp => kvp.Value).FirstOrDefault(item => item.CopyType == CopyToOutputDirectoryType.CopyAlways);
-            if (copyAlwaysItem.Path != null)
+            string copyAlwaysItemPath = _items.SelectMany(kvp => kvp.Value).FirstOrDefault(item => item.CopyType == CopyToOutputDirectoryType.CopyAlways).Path;
+
+            if (copyAlwaysItemPath != null)
             {
-                return Fail(logger, $"Item '{copyAlwaysItem.Path}' has CopyToOutputDirectory set to 'Always', not up to date.", "CopyAlwaysItemExists");
+                return Fail(logger, $"Item '{copyAlwaysItemPath}' has CopyToOutputDirectory set to 'CopyAlways', not up to date.", "CopyAlwaysItemExists");
```
[the remark this answers is not in the document]
Comment (drewnoakes) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UpToDate/BuildUpToDateCheck.cs]: Reverted to `Always` in e3794f5265020e7512. I was reflecting the enum member name. Matching the XML value makes more sense.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UpToDate/BuildUpToDateCheck.cs [hunk 3]
```diff
@@ -358,33 +316,88 @@ private bool CheckGlobalConditions(BuildAction buildAction, BuildUpToDateCheckLo
 
         private IEnumerable<string> CollectInputs(BuildUpToDateCheckLogger logger)
         {
-            var inputs = new HashSet<string>(StringComparers.Paths);
-
             logger.Verbose("Adding project file inputs:");
```
[the remark this answers is not in the document]
Comment (drewnoakes) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UpToDate/BuildUpToDateCheck.cs]: The consumer of this method (`IsUpToDateAsync`) performs its own de-duping via the `timestampCache`. The outcome is the same, but we avoid allocating the collection and hashing each input twice. The main reason I did this was to allow breaking early from the input scan as soon as an input is too new.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UpToDate/BuildUpToDateCheck.cs (hunk 1 above)
Comment (drewnoakes) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UpToDate/BuildUpToDateCheck.cs]: Once I'd convinced myself it was safe, it felt easier to read to me too. Unfortunately other optimisations made things less elegant in other places.

**Comment by drewnoakes** [comment]
> I appreciate the concise split of commits

Great. I like that style too. Some projects prefer chunkier commits.

> Does this make much of an impact on GC usage? I would expect see some Gen1 improvements here over say something like Roslyn.sln, both opening the solution & switching between a branch with a large amount of changes. I can help you investigate that - don't block merge on it though.

I suspect we'd see more traffic with less promotion from Gen0, but until profiled it's only a guess. I would be very happy to hear suggestions on approaches for profiling. I tried the built in profiler on Roslyn.sln but it ran out of memory. I'll give dotMemory a shot next.

I've been looking into adding tests for `BuildUpToDateCheck` today, which has required skilling up on a few things. Am happy to hold off on merging until tests are in and it's profiled to verify improvement. Will this block anyone or likely conflict with other near-term changes?

**Comment by davkean** [comment]
@drewnoakes Let's sync up next week and I'll walk your through PerfView, that's what we tend to use. The UI is a bit clunky and hard to understand, but it's pretty decent once you get a handle on it. If you try today, do the following:

1. Use 64-bit version of it
2. Collect -> Collect
3. Turn off merging/zipping to speed up investigations
4. Up the buffer significantly to many GB
5. Start record and repro your scenario, the smaller it is the better. I'll teach you next week how to use events ("Finish solution load") to filter the trace to the same scenario exactly
6. Use combination of `GC Heap Alloc Ignore Free`/`GCStats` to see how much improvement you've made.

**Comment by drewnoakes** [comment]
> Let's sync up next week

Sounds good. I'll spend some time today familiarising myself with it to make things more productive next week. I can get online earlier so we have more overlap.

Thanks for the notes. I'll start with those. Won't have access to `symweb` until I get a machine on the VPN, which is hopefully happening today.

**Comment by davkean** [comment]
@dotnet-bot retest windows_integration_debug please

It ran into: https://github.com/dotnet/project-system/issues/3924.

**Comment by davkean** [comment]
@dotnet-bot retest windows_integration_debug please

**Comment by davkean** [comment]
@dotnet-bot retest this please

**Comment by davkean** [comment]
@drewnoakes I'd merged this in without blocking on the tests.

**Comment by drewnoakes** [comment]
Sure thing.

**Comment by lifengl** [comment]
Cool, thanks a lot to take care of that quickly!

**Pull request merged**